### PR TITLE
Add --enable-border-router configure option

### PR DESCRIPTION
**Pull request by chshu**
Add --enable-border-router configure option
* Use `OPENTHREAD_ENABLE_BORDER_ROUTER` feature flag to wrap Border Router related features
* MTD devices could also act as a Border Router

For the devices which don't enable Border Router, the code size will reduce more than 3K bytes:
```
$ make -f examples/Makefile-cc2650
$ arm-none-eabi-size output/cc2650/bin/ot-cli-mtd 
   text	   data	    bss	    dec	    hex	filename
  85564	    736	  12152	  98452	  18094	output/cc2650/bin/ot-cli-mtd

$ make -f examples/Makefile-cc2650 BORDER_ROUTER=1
$ arm-none-eabi-size output/cc2650/bin/ot-cli-mtd 
   text	   data	    bss	    dec	    hex	filename
  88804	    736	  12424	 101964	  18e4c	output/cc2650/bin/ot-cli-mtd
```

### CODE REVIEW

**Comment by abtink** [comment]
This is great. Thanks @chshu  . Particularly the reduced code size sounds great.
Brain-storming with @jwhui  on how to name the modules/methods, we came up with a proposal/suggestion:

We can define a new set of APIs all starting with `otBorderRouter` for all border router related
OT public functions. Can then add them in a new header file `border_router.h` and their corresponding implementation in `border_router_api.cpp`.  Both of these tied to `OPENTHREAD_ENABLE_BORDER_ROUTER` feature flag).

We can include the following APIs in the new border router module:

- `otBorderRoutetGetLocalNetData()` replacing `otNetDataGetLocal()`
- `otBorderRouterAddOnMeshPrefix()` replacing `otNetDataAddPrefixInfo()` (changes the local network data)
- `otBorderRouterRemoveOnMeshPrefix()` replacing `otNetDataRemovePrefixInfo()` (changes the local network data)
- `otBorderRouterAddRoute()` replacing `otNetDataAddRoute()`.
- `otBorderRouterRemoveRoute()` replacing `otNetDataRemoveRoute()`.
- `otBorderRouterRegisterNetData()` replacing `otNetDataRegister()`.

Two new methods to iterate through local network data on-mesh-prefix and route-prefix
- `otBorderRouterGetNextLocalOnMeshPrefix()`
- `otBorderRouterGetNextLocalRoute()`

We then change the `netdata.h` methods to provide the leader's (partition's) network data only: 

- `otNetDataGet()` to replace `otNetDataGetLeader()`
- `otNetDataGetNextOnMeshPrefix()` to replace `otNetDataGetNextPrefixInfo()` without having `aLocal` parameter and always using the leader's net data.
- `otNetDataGetNextRoute()` to replace `otNetDataGetNextRoute()` without having `aLocal` parameter and always using leader's net data.
- `otNetDataGetVersion()` (as is)
- `otNetDataGetStableVersion()` (as is)

Thoughts?

**Comment by chshu** [comment]
@abtink, The proposal sounds great, I will update the PR accordingly (I think the changes are mainly focused on naming updates, without any code logic update).

**Comment by abtink** [comment]
Yes. Thanks @chshu

**Comment by chshu** [comment]
@abtink @jwhui , The PR has been updated according to the border router naming proposal, please have a review, thanks.

File: include/openthread/border_router.h [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+ * @param[in]  aInstance A pointer to an OpenThread instance.
+ *
+ * retval OT_ERROR_NONE  Successfully queued a Server Data Request message for delivery.
+ *
+ * @sa otBorderRouterAddOnMeshPrefix
+ * @sa otBorderRouterRemoveOnMeshPrefix
+ * @sa otBorderRouterAddRoute
+ * @sa otBorderRouterRemoveRoute
+ */
+OTAPI otError OTCALL otBorderRouterRegisterNetData(otInstance *aInstance);
```
Comment (jwhui) [inline include/openthread/border_router.h]: How about simply `otBorderRouterRegister(...)`?

File: include/openthread/border_router.h [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+ * Add a border router configuration to the local network data.
+ *
+ * @param[in]  aInstance A pointer to an OpenThread instance.
+ * @param[in]  aConfig   A pointer to the border router configuration.
+ *
+ * @retval OT_ERROR_NONE          Successfully added the configuration to the local network data.
+ * @retval OT_ERROR_INVALID_ARGS  One or more configuration parameters were invalid.
+ * @retval OT_ERROR_NO_BUFS       Not enough room is available to add the configuration to the local network data.
+ *
+ * @sa otRemoveBorderRouter
```
Comment (jwhui) [inline include/openthread/border_router.h]: `@sa` tags need to be updated here and below

File: include/openthread/border_router.h [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+ * @param[in]     aInstance  A pointer to an OpenThread instance.
+ * @param[inout]  aIterator  A pointer to the Network Data iterator context. To get the first on-mesh entry
+                             it should be set to OT_NETWORK_DATA_ITERATOR_INIT.
+ * @param[out]    aConfig    A pointer to where the On Mesh Prefix information will be placed.
+ *
+ * @retval OT_ERROR_NONE       Successfully found the next On Mesh prefix.
+ * @retval OT_ERROR_NOT_FOUND  No subsequent On Mesh prefix exists in the Thread Network Data.
+ *
+ */
+OTAPI otError OTCALL otBorderRouterGetNextLocalOnMeshPrefix(otInstance *aInstance, otNetworkDataIterator *aIterator,
```
Comment (jwhui) [inline include/openthread/border_router.h]: I think the `otBorderRouter*` module should be limited to modifying local network data and registering that local network data with the leader.  If that is the case, then we can remove `Local` from the `otBorderRouter*` calls.  In this case, we can simplify to `otBorderRouterGetNextOnMeshPrefix(...)`.  Thoughts?

File: include/openthread/border_router.h (hunk 1 above)
Comment (chshu) [inline include/openthread/border_router.h]: Done, thanks.

File: include/openthread/border_router.h (hunk 3 above)
Comment (chshu) [inline include/openthread/border_router.h]: Yes, it does make sense. I have removed `Local` from `otBorderRouter*` calls.

File: include/openthread/border_router.h (hunk 2 above)
Comment (chshu) [inline include/openthread/border_router.h]: Fixed, thanks.

**Pull request merged**